src/coreclr/jit/valuenum.cpp

```diff
@@ -14998,3 +14998,39 @@ CORINFO_CLASS_HANDLE ValueNumStore::GetObjectType(ValueNum vn, bool* pIsExact, b
 
     return NO_CLASS_HANDLE;
 }
+
+//--------------------------------------------------------------------------------
+// PeelOffsets: Peel all additions with a constant offset away from the
+// specified VN.
+//
+// Arguments:
+//    vn     - [in, out] The VN. Will be modified to the base VN that the offsets are added to.
+//    offset - [out] The offsets peeled out of the VNF_ADD funcs.
+//
+void ValueNumStore::PeelOffsets(ValueNum* vn, target_ssize_t* offset)
+{
+#ifdef DEBUG
+    var_types vnType = TypeOfVN(*vn);
+    assert((vnType == TYP_I_IMPL) || (vnType == TYP_REF) || (vnType == TYP_BYREF));
+#endif
+
+    *offset = 0;
+    VNFuncApp app;
+    while (GetVNFunc(*vn, &app) && (app.m_func == VNF_ADD))
```

Comment:
Seems like VN ought to commute so that constants are (say) rightmost / and or fold chains of constant adds... but I take it you're seeing add chains or mixtures of constants left and right.

Reply:
To be honest I haven't checked whether we see cases that require the extra unwrapping... I just wrote this one fully generally to match `gtPeelOffsets` and `Scev::PeelAdditions`.
We do canonicalize commutative operators, but only in terms of raw VN number, keeping the smallest VN number as op1. Definitely seems like we could make sure constants go to the right.

Reply:
> Definitely seems like we could make sure constants go to the right.

I remember I tried that but found no diffs